### PR TITLE
Added filtering on empty values.

### PR DESCRIPTION
**Pull request by areksolek**
Added filtering on empty values.

### CODE REVIEW

File: layouts/basic/modules/Vtiger/List/Field/Base.tpl [hunk 1]
```diff
@@ -18,6 +18,7 @@
 {else}
 	{assign var=SEARCH_VALUE value=''}
 {/if}
+{assign var=READONLY value=$FIELD_MODEL->getReadOnlyFieldInSearchParams($CONSTANT_READONLY, $TEMPORARILY_READONLY)}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/List/Field/Base.tpl]: ?

File: layouts/basic/modules/Vtiger/ListViewContents.tpl [hunk 2]
```diff
@@ -24,6 +24,7 @@
 	<input type="hidden" id="excludedIds" name="excludedIds"/>
 	<input type='hidden' value="{$PAGING_MODEL->getPageLimit()}" id='pageLimit'>
 	<input type="hidden" value="{$LISTVIEW_ENTRIES_COUNT}" id="noOfEntries">
+	<input type="hidden" class="js-temporarily-readonly" data-js="value" value='{$TEMPORARILY_READONLY}'/>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ListViewContents.tpl]: ?

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

File: layouts/basic/modules/Vtiger/ListViewContents.tpl [hunk 3]
```diff
@@ -24,7 +24,7 @@
 	<input type="hidden" id="excludedIds" name="excludedIds"/>
 	<input type='hidden' value="{$PAGING_MODEL->getPageLimit()}" id='pageLimit'>
 	<input type="hidden" value="{$LISTVIEW_ENTRIES_COUNT}" id="noOfEntries">
-	<input type="hidden" class="js-temporarily-readonly" data-js="value" value='{$TEMPORARILY_READONLY}'/>
+	<input type="hidden" class="js-locked-fields" data-js="value" value='{$SEARCH_COLUMN_EMPTI_CONDITION}'/>
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/ListViewContents.tpl]: SEARCH_COLUMN_EMPT**Y**_CONDITION

File: modules/Vtiger/models/Field.php [hunk 4]
```diff
@@ -1553,6 +1532,19 @@ public function isActiveSearchView()
...
+	 * Empty value search in view.
+	 *
+	 * @return bool
+	 */
+	public function emptyValueSearchInView(): bool
+	{
+		if ($this->get('disabledField')) {
+			return false;
+		}
+		return true;
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Field.php]: return empty($this->get('disabledField'));

File: modules/Vtiger/models/ListView.php [hunk 5]
```diff
@@ -613,4 +613,41 @@ public function getListViewCount()
...
+	 *
+	 * @return void
+	 */
+	public function fieldsPermanentlyBlocked(): void
+	{
+		$moduleModel = $this->getModule()->getFields();
+		foreach ($this->getArray('search_params') as $values) {
+			if (\is_array($values)) {
+				foreach ($values as $value) {
+					$fieldModel = $moduleModel[$value['field_name']];
```
Comment (rskrzypczak) [inline modules/Vtiger/models/ListView.php]: $moduleModel->getFieldByName()

File: modules/Vtiger/models/RelationListView.php [hunk 6]
```diff
@@ -695,4 +695,41 @@ public function isWidgetsList(): bool
...
+	 *
+	 * @param App\Request $request
+	 *
+	 * @return void
+	 */
+	public function lockedFields(App\Request $request): void
+	{
+		$moduleModel = $this->getRelationModel()->getRelationModuleModel()->getFields();
+		if (!$request->isEmpty('fieldsLocked')) {
+			foreach ($request->getArray('fieldsLocked', 'AlnumExtended') as $value) {
```
Comment (rskrzypczak) [inline modules/Vtiger/models/RelationListView.php]: AlnumExtended ?

File: modules/Vtiger/models/ListView.php [hunk 7]
```diff
@@ -613,4 +613,41 @@ public function getListViewCount()
...
+	 * @return void
+	 */
+	public function lockedFields(App\Request $request): void
+	{
+		$moduleModel = $this->getModule()->getFields();
+		if (!$request->isEmpty('fieldsLocked')) {
+			foreach ($request->getArray('fieldsLocked', 'AlnumExtended') as $value) {
+				$fieldModel = $moduleModel[$value];
+				$fieldModel->set('disabledField', true);
+				$fieldModel->set('fromOutsideList', false);
```
Comment (rskrzypczak) [inline modules/Vtiger/models/ListView.php]: ?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: modules/Vtiger/models/Field.php [hunk 8]
```diff
@@ -1532,6 +1532,16 @@ public function isActiveSearchView()
...
 	}
 
+	/**
+	 * Empty value search in view.
+	 *
+	 * @return bool
+	 */
+	public function emptyValueSearchInView(): bool
+	{
+		return empty($this->get('disabledField'));
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/Field.php]:
```suggestion
		return empty($this->get('searchlockedFields'));
```

File: modules/Vtiger/models/ListView.php [hunk 9]
```diff
@@ -613,4 +613,41 @@ public function getListViewCount()
...
+
+	/**
+	 * Fields permanently blocked if found in search value.
+	 *
+	 * @return void
+	 */
+	public function fieldsPermanentlyBlocked(): void
+	{
+		$moduleModel = $this->getModule();
+		foreach ($this->getArray('search_params') as $values) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/ListView.php:625]: ?

File: modules/Vtiger/models/ListView.php [hunk 10]
```diff
@@ -613,4 +613,41 @@ public function getListViewCount()
 		$this->loadListViewCondition();
 		return $this->getQueryGenerator()->createQuery()->count();
 	}
+
+	/**
+	 * Fields permanently blocked if found in search value.
+	 *
+	 * @return void
+	 */
+	public function fieldsPermanentlyBlocked(): void
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/ListView.php]:
```suggestion
	public function loadSearchLockedFields(): void
```

Comment (mariuszkrzaczkowski) [reply modules/Vtiger/models/ListView.php]: ?

File: modules/Vtiger/models/ListView.php (hunk 7 above)
Comment (mariuszkrzaczkowski) [reply modules/Vtiger/models/ListView.php]: ?

File: modules/Vtiger/models/ListView.php [hunk 11]
```diff
@@ -613,4 +613,41 @@ public function getListViewCount()
...
+	{
+		$moduleModel = $this->getModule();
+		foreach ($this->getArray('search_params') as $values) {
+			if (\is_array($values)) {
+				foreach ($values as $value) {
+					$fieldModel = $moduleModel->getFieldByName($value['field_name']);
+					$fieldModel->set('fromOutsideList', true);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/ListView.php]: ?

File: modules/Vtiger/views/List.php [hunk 12]
```diff
@@ -233,6 +233,8 @@ public function initializeListViewContents(App\Request $request, Vtiger_Viewer $
 		if (!empty($searchParams) && \is_array($searchParams)) {
 			$transformedSearchParams = $this->listViewModel->getQueryGenerator()->parseBaseSearchParamsToCondition($searchParams);
 			$this->listViewModel->set('search_params', $transformedSearchParams);
+			$this->listViewModel->fieldsPermanentlyBlocked();
+			$this->listViewModel->lockedFields($request);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/List.php]: ?

File: modules/Vtiger/views/List.php [hunk 13]
```diff
@@ -262,6 +264,7 @@ public function initializeListViewContents(App\Request $request, Vtiger_Viewer $
 		$viewer->assign('PAGING_MODEL', $pagingModel);
 		$viewer->assign('PAGE_NUMBER', $pageNumber);
 		$viewer->assign('ORDER_BY', $orderBy);
+		$viewer->assign('SEARCH_COLUMN_EMPTY_CONDITION', $request->isEmpty('fieldsLocked', true) ? false : \App\Json::encode($request->getArray('fieldsLocked')));
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/List.php]: ?

Comment (mariuszkrzaczkowski) [reply modules/Vtiger/views/List.php]: lockedFields

File: public_html/layouts/basic/modules/Vtiger/resources/ListSearch.js [hunk 14]
```diff
@@ -90,6 +91,95 @@ jQuery.Class(
...
+			return {
+				search_params: [newParams],
+				fieldsLocked: temporarily
+			};
+		},
+
+		/**
+		 * Register list search if value empty.
+		 */
+		registerListSearchIfEmptyValue: function () {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ListSearch.js]:
```suggestion
		registerListSearchEmptyValue: function () {
```

File: public_html/layouts/basic/modules/Vtiger/resources/ListSearch.js [hunk 15]
```diff
@@ -90,6 +91,95 @@ jQuery.Class(
...
+			let listViewContainer = this.getContainer();
+			let thisInstance = this;
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ListSearch.js]:
```suggestion
			const self = this;
```

File: public_html/layouts/basic/modules/Vtiger/resources/RecordsList.js [hunk 16]
```diff
@@ -48,11 +48,18 @@ $.Class(
 				totalCount: this.container.find('.js-total-count').val(),
 				noOfEntries: this.container.find('.js-no-entries').val(),
 				filterFields: JSON.parse(this.container.find('.js-filter-fields').val()),
+				fieldsLocked: this.container.find('.js-locked-fields').val(),
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/RecordsList.js]: lockedFields

File: public_html/layouts/basic/modules/Vtiger/resources/RecordsList.js [hunk 17]
```diff
@@ -48,11 +48,18 @@ $.Class(
...
 				onlyBody: true,
 				cvId: this.getFilterSelectElement().val()
 			};
 			let searchValue = this.listSearchInstance.getAlphabetSearchValue();
-			params['search_params'] = JSON.stringify(this.listSearchInstance.getListSearchParams(true));
+			let searchParams = JSON.parse(this.container.find('#search_params').val())[0];
+			if (searchParams !== undefined && searchParams.length > 0) {
+				params['search_params'] = this.container.find('#search_params').val();
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/RecordsList.js:58]: ?

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

File: modules/Vtiger/models/Field.php [hunk 18]
```diff
@@ -1532,6 +1532,16 @@ public function isActiveSearchView()
 		return $this->getUITypeModel()->isActiveSearchView();
 	}
 
+	/**
+	 * Empty value search in view.
+	 *
+	 * @return bool
+	 */
+	public function emptyValueSearchInView(): bool
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/Field.php]:
```suggestion
	public function searchlockedFields(): bool
```

File: layouts/basic/modules/Vtiger/List/Field/Base.tpl [hunk 19]
```diff
@@ -34,8 +34,15 @@
 			</div>
 		</div>
 	{else}
-		<input type="text" name="{$FIELD_MODEL->getName()}" {if !empty($FIELD_MODEL->get('source_field_name'))} data-source-field-name="{$FIELD_MODEL->get('source_field_name')}" data-module-name="{$FIELD_MODEL->getModuleName()}"
-		{/if} class="listSearchContributor form-control" value="{$SEARCH_VALUE}" title='{$LABEL['label']}' data-fieldinfo='{$FIELD_INFO|escape}' {if !$FIELD_MODEL->isActiveSearchView()}disabled{/if}/>
+		<div class="input-group">
+			<input type="text" name="{$FIELD_MODEL->getName()}"  {if !empty($FIELD_MODEL->get('source_field_name'))} data-source-field-name="{$FIELD_MODEL->get('source_field_name')}" data-module-name="{$FIELD_MODEL->getModuleName()}"
+			{/if} class="listSearchContributor form-control" value="{$SEARCH_VALUE}" title='{$LABEL['label']}' data-fieldinfo='{$FIELD_INFO|escape}' {if !$FIELD_MODEL->searchlockedFields() || !$FIELD_MODEL->isActiveSearchView()}disabled{/if}/>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/List/Field/Base.tpl]: searchLockedFields

File: modules/Vtiger/models/Field.php [hunk 20]
```diff
@@ -1526,12 +1526,22 @@ public function getMaxValue(): ?string
...
 	}
 
+	/**
+	 * Empty value search in view.
+	 *
+	 * @return bool
+	 */
+	public function searchlockedFields(): bool
+	{
+		return empty($this->get('searchlockedFields'));
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/Field.php]: searchLockedFields

File: modules/Vtiger/models/Field.php [hunk 21]
```diff
@@ -1526,12 +1526,22 @@ public function getMaxValue(): ?string
 
 	public function isActiveSearchView()
 	{
-		if ($this->get('fromOutsideList')) {
+		if ($this->get('fromOutsideList') || $this->get('fieldsPermanentlyBlocked')) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/Field.php]: ?

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

File: public_html/layouts/basic/modules/Vtiger/resources/ListSearch.js [hunk 22]
```diff
@@ -40,6 +40,10 @@ jQuery.Class(
 		container: false,
 		reletedInstance: false,
 		viewName: false,
+		/**
+		 * Column name from last search.
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/ListSearch.js:44]: /**
 * @type {string}
 */

Comment (mariuszkrzaczkowski) [reply public_html/layouts/basic/modules/Vtiger/resources/ListSearch.js:44]: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

**Pull request merged**